docs: update router containerization and deployment content

shorgi opened the pull request:
Updated and reorganized Containerization content, primarily rewriting Kubernetes topic as quickstarts and short examples to deploying router on k8s with Helm.

Previews: 
- [Containerization overview](https://deploy-preview-3943--apollo-router-docs.netlify.app/containerization/overview)
- [Deploy in Kubernetes](https://deploy-preview-3943--apollo-router-docs.netlify.app/containerization/kubernetes) 
- [Run in Docker](https://deploy-preview-3943--apollo-router-docs.netlify.app/containerization/docker)

garypen commented on docs/source/containerization/kubernetes.mdx:
```diff
@@ -1,246 +1,274 @@
+
+> For example, the Helm chart of the latest release of the router is available [here](https://github.com/apollographql/router/tree/v1.31.1/helm/chart/router).
+
+Since the router v0.14.0, Apollo has also released a Helm chart for the router as an [Open Container Initiative (OCI)](https://helm.sh/docs/topics/registries/) image in the GitHub Container registry. 
```
We prefer people to use the `OCI` release chart rather than using the version in the repo. We should probably stop documenting that altogether.

garypen replied:
```suggestion
Since the router v0.14.0, Apollo has also released a Helm chart for the router as an [Open Container Initiative (OCI)](https://helm.sh/docs/topics/registries/) image in our GitHub Container registry. 
```

garypen commented on docs/source/containerization/kubernetes.mdx:
```diff
@@ -1,246 +1,274 @@
+
+> The path to the OCI router chart is `oci://ghcr.io/apollographql/helm-charts/router`.
+
+<ExpansionPanel title="Click to expand example chart for router v1.31.0">
```
```suggestion
<ExpansionPanel title="Click to expand example values for router v1.31.0">
```

garypen commented on docs/source/containerization/kubernetes.mdx:
```diff
@@ -1,246 +1,274 @@
+
+Configuring a `values.yaml` file for settings common across environments.
+
+<ExpansionPanel title="Example router Helm chart for common configuration across environments">
```
```suggestion
<ExpansionPanel title="Example router Helm values for common configuration across environments">
```

garypen commented on docs/source/containerization/kubernetes.mdx:
```diff
@@ -1,246 +1,274 @@
+
+## Install router Helm chart
+
+You can install the router chart from either a local repository or the container registry.
```
```suggestion
You can install the router chart from either a local repository or the Apollo container registry.
```
also: Maybe repeat location here (`oci://ghcr... etc...)

garypen commented on docs/source/containerization/kubernetes.mdx:
````diff
@@ -1,246 +1,274 @@
+Setting router command-line arguments and YAML config file keys/values. 
+
+```yaml title="values.yaml"
+router:
+  router:
````
Everywhere where you have:
```
router:
  router:
```
The first `router:` needs to be deleted and all the rest of the text needs to be left shifted by 2 bytes.

REASON: We are using an in-house chart which "wraps" the main router chart. We do that to make it simple to deploy `Rhai` and configure `istio`. I don't think we are committing to documenting that here, since it's Apollo specific, but if we are doing that, then we need to decide how to document it.

garypen commented on docs/source/containerization/kubernetes.mdx:
````diff
@@ -1,246 +1,274 @@
+          timeout: 5min
+```
+
+Support for `@defer` is unsupported in gateway, so it needs to be disabled in the router.
+
+```yaml title="values.yaml"
+router:
+  router:
+    configuration:
+      supergraph:
+        defer_support: false
````
I'm not sure if we should document this. We've enabled support for defer now.

garypen commented on docs/source/containerization/kubernetes.mdx:
````diff
@@ -1,246 +1,274 @@
+* Install the OCI image from the GiHub container registry:
+
+  ```bash title="OCI registry"
+  helm install --set router.configuration.telemetry.metrics.prometheus.enabled=true --set managedFederation.apiKey="<graph-api-key>" --set managedFederation.graphRef="<graph-ref>" --create-namespace --namespace router-deploy router-test oci://ghcr.io/apollographql/helm-charts/router --version 1.0.0-rc.8 --values router/values.yaml
````
This needs some refinement. e.g.: prometheus we don't want to set and the multiple values files for common and env are missing.

garypen commented on docs/source/containerization/kubernetes.mdx:
````diff
@@ -1,246 +1,274 @@
+  helm install --set router.configuration.telemetry.metrics.prometheus.enabled=true --set managedFederation.apiKey="<graph-api-key>" --set managedFederation.graphRef="<graph-ref>" --create-namespace --namespace router-deploy router-test oci://ghcr.io/apollographql/helm-charts/router --version 1.0.0-rc.8 --values router/values.yaml
+  ```
+
+### Install Helm chart from local repository
+
+* Clone the [router repository](https://github.com/apollographql/router).
+
+* Change to the directory in your local repo with the router chart:
+
+  ```bash
+  cd <repo-directory>/helm/chart
+  ```
+
+* Install the local chart, `router`:
+
+  ```bash title="Local repository"
+  helm install --set router.configuration.telemetry.metrics.prometheus.enabled=true --set managedFederation.apiKey="<graph-api-key>" --set managedFederation.graphRef="<graph-ref>" --create-namespace --namespace router-deploy router-test router --values router/values.yaml
+  ```
+
+  (TODO: clean up and describe the used helm install flags)
+
+  ```
+  In both the following examples, we are using helm to install the router:
+  - into namespace "router-deploy" (create namespace if it doesn't exist)
+  - with helm install name "router-test"
+  - with support for prometheus enabled
+  ```
+
````
I don't think we should document how to do this. It's better to use our OCI published chart.

garypen reviewed (commented):
I've left some comments mainly for discussion when we meet later.

shorgi commented on docs/source/containerization/kubernetes.mdx:
````diff
@@ -1,264 +1,249 @@
-      command: ['nc']
-      args: ['-vz','-w','1','release-name-router:80']
-  restartPolicy: Never
+* Create a YAML file, `my_values.yaml`, to contain additional values that override default values.
+* Edit the file with values to enable metrics endpoints:
+
+  ```yaml title="my_values.yaml"
+  router:
+    configuration:
+      telemetry:
+        metrics:
+          prometheus:
+            enabled: true
+            listen: 0.0.0.0:9090
+            path: "/metrics"
+          otlp:
+            temporality: delta
+            endpoint: <otlp-endpoint-addr>
+  ```
+
+  * `router.telemetry.metrics.prometheus` was already configured but disabled (`enabled: false`) by default. This configuration sets `enabled: true`.
+  * `router.telemetry.metrics.otlp` is enabled by inclusion.
+
+* Deploy the router with the additional YAML configuration file. For example, starting with the `helm install` command from the basic deployment step, append `--values my_values.yaml`:
+
+  ```bash
+  helm install --set managedFederation.apiKey="<graph-api-key>" --set managedFederation.graphRef="<graph-ref>"  oci://ghcr.io/apollographql/helm-charts/router --version <router-version> --values router/values.yaml --namespace <router-namespace> --values my_values.yaml
````
@garypen - are the steps here correct, from creating a new values YAML, editing it, and appending it to the `helm install` deployment command? (Those general steps are repeated below for Rhai scripting and the coprocessor)

garypen replied:
It is correct. I'd just make one cosmetic change to remove the `--namespace` entry from between the two values files. So:
```
helm install --namespace <router-namespace> --set managedFederation.apiKey="<graph-api-key>" --set managedFederation.graphRef="<graph-ref>"  oci://ghcr.io/apollographql/helm-charts/router --version <router-version> --values router/values.yaml --values my_values.yaml
```

garypen commented on docs/source/containerization/kubernetes.mdx:
```diff
@@ -1,264 +1,270 @@
+When [migrating from `@apollo/gateway` to the Apollo Router](../migrating-from-gateway), consider the following tips to maximize the compatibility of your router deployment.
 
-## Using `istio` with the router
+* The gateway's maximum supported request size is 20MB, while the router's default maximum request size is less. For compatibility, set the deployed router's maximium request size to 20MB:
```
I'm not sure about this. I think it's just that Apollo's gateway was configured to support 20MB. The router by default has a max of 2Mb, so we increased it to 20MB.

This is more general wording that will work for all users:

By default the router has a maximum supported request size of 2MB. If your gateway accepts requests larger than 2MB, you can use the following configuration to ensure that the router is compatible with your gateway deployment.

<yaml extract>

garypen commented on docs/source/containerization/kubernetes.mdx:
```diff
@@ -1,264 +1,270 @@
+      router:
+        timeout: 6min
+      all:
+        experimental_enable_http2: false
```
Remove line 259. We aren't using that in-house. It was a short lived experiment.

garypen commented on docs/source/containerization/kubernetes.mdx:
````diff
@@ -1,264 +1,270 @@
+  helm install --namespace <router-namespace> --set managedFederation.apiKey="<graph-api-key>" --set managedFederation.graphRef="<graph-ref>"  oci://ghcr.io/apollographql/helm-charts/router --version <router-version> --values router/values.yaml --values my_values.yaml
+  ```
+
+## Deploy with Rhai scripts
+
+The router supports [Rhai scripting](../../customizations/rhai) to add custom functionality.
+
+To enable Rhai scripts in your deployed router through a YAML configuration file:
+
+* Create a YAML file, `my_values.yaml`, to contain additional values that override default values.
+* Edit the file with values to enable Rhai scripts:
+
+  ```yaml title="my_values.yaml"
+  router:
+    configuration:
+      rhai:
+        scripts: <rhai-scripts-path>
+        main: <rhai-main-script>
+  ```
+
+* Deploy the router with the additional YAML configuration file. For example, starting with the `helm install` command from the basic deployment step, append `--values my_values.yaml`:
+
+  ```bash
+  helm install  --namespace <router-namespace> --set managedFederation.apiKey="<graph-api-key>" --set managedFederation.graphRef="<graph-ref>"  oci://ghcr.io/apollographql/helm-charts/router --version <router-version> --values router/values.yaml --values my_values.yaml
+  ```
````
One alternative: I [already created an example repo](https://github.com/apollographql/in-house-router-example) of how to approach this problem which I have shared with customers in the past. It's a really stripped down version of how we deploy our Rhai scripts in-house.
Can we write some text in this section which explains why this is a more complex problem than the others and then links to [this repo](https://github.com/apollographql/in-house-router-example) to provide a fully worked example?

shorgi commented on docs/source/containerization/kubernetes.mdx:
```diff
@@ -1,264 +1,270 @@
+    url: http://127.0.0.1:8081 # Replace with the URL of your coprocessor's HTTP endpoint.
+    router:
+      request:
+        headers: false
```
Changing to show the typical coproc config

shorgi commented on docs/source/containerization/kubernetes.mdx:
````diff
@@ -1,264 +1,270 @@
+
+> ⚠️ Your Kubernetes version must be compatible with Helm v3. For details, see [Helm Version Support Policy](https://helm.sh/docs/topics/version_skew/#supported-version-skew).
+
+* Log in with Helm to the Apollo container registry in GitHub.
+  
+  * Get a GitHub OCI token and save it in an environment variable, `GITHUB_OCI_TOKEN`. For reference, follow the guide for [authenticating to the GitHub container registry](https://docs.github.com/en/packages/working-with-a-github-packages-registry/working-with-the-container-registry#authenticating-to-the-container-registry).
+
+  * Log in with the `helm registry login` command, using your saved `GITHUB_OCI_TOKEN` and GitHub username:
+
+    ```bash
+    echo ${GITHUB_OCI_TOKEN} | helm registry login -u <username> --password-stdin ghcr.io
+    ```
+
+* After logging in, verify your access to the registry by showing the latest router chart values with the `helm show values` command:
````
I needed to login (using these steps) to run `helm show values oci://ghcr.io/apollographql/helm-charts/router`, and to verify I tried it just now on my personal laptop and got a 403:

> Error: failed to authorize: failed to fetch anonymous token: unexpected status from GET request to https://ghcr.io/token?scope=repository%3Aapollographql%2Fhelm-charts%2Frouter%3Apull&scope=repository%3Auser%2Fimage%3Apull&service=ghcr.io: 403 Forbidden

Is this unexpected response from the registry?

garypen replied:
Sorry, you are correct. You do need to login. I thought they had removed that requirement, but I just checked from a machine where I knew I was definitely not logged in and confirmed I couldn't access the chart.

Meschreiber commented on docs/shared/helm-show-router-output.mdx:
```diff
@@ -0,0 +1,238 @@
+managedFederation:
+  # -- If using managed federation, the graph API key to identify router to Studio
+  apiKey:
+  # -- If using managed federation, use existing Secret which stores the graph API key instead of creating a new one.
```
```suggestion
  # -- If using managed federation, use an existing secret which stores the graph API key instead of creating a new one.
```

shorgi replied:
@garypen - a note to incorporate the copyedits from this PR into both the release and in-house router charts

Meschreiber commented on docs/shared/router-common-config.mdx:
```diff
@@ -0,0 +1,210 @@
+# This is a YAML-formatted file.
+# Declare variables to be passed into your templates.
+
+# Anything we set here applies to the in-house-router chart, and 
```
```suggestion
# Anything set here applies to the in-house-router chart, and 
```

Meschreiber commented on docs/shared/router-common-config.mdx:
```diff
@@ -0,0 +1,210 @@
+# Declare variables to be passed into your templates.
+
+# Anything we set here applies to the in-house-router chart, and 
+# all configuration that we set here is pushed to all sub-charts
```
```suggestion
# all configuration that set here is pushed to all sub-charts
```

Meschreiber commented on docs/shared/router-common-config.mdx:
```diff
@@ -0,0 +1,210 @@
+
+# Anything we set here applies to the in-house-router chart, and 
+# all configuration that we set here is pushed to all sub-charts
+# (because of the way Helm works:
```
```suggestion
# because of the way Helm works:
```

Meschreiber commented on docs/shared/router-common-config.mdx:
```diff
@@ -0,0 +1,210 @@
+# Anything we set here applies to the in-house-router chart, and 
+# all configuration that we set here is pushed to all sub-charts
+# (because of the way Helm works:
+#  https://helm.sh/docs/chart_template_guide/subcharts_and_globals/)
```
```suggestion
#  https://helm.sh/docs/chart_template_guide/subcharts_and_globals/
```

Meschreiber commented on docs/shared/router-common-config.mdx:
```diff
@@ -0,0 +1,210 @@
+          timeout: 5min
+      supergraph:
+        introspection: true
+        # We are intentionally DISABLING DEFER SUPPORT in the Router until
```
```suggestion
        #  this intentionally disables `@defer` support in the router until
```

Meschreiber commented on docs/shared/router-common-config.mdx:
```diff
@@ -0,0 +1,210 @@
+      supergraph:
+        introspection: true
+        # We are intentionally DISABLING DEFER SUPPORT in the Router until
+        # we are out of our transition period from the Gateway to the Router.
```
```suggestion
        # until the transition from the gateway to the router is complete.
```

Meschreiber commented on docs/shared/router-common-config.mdx:
```diff
@@ -0,0 +1,210 @@
+
+router:
+  serviceAccount:
+    # We do that ourselves in permissions.yaml
```
```suggestion
    # The service account is created permissions.yaml
```

Meschreiber commented on docs/shared/router-common-config.mdx:
```diff
@@ -0,0 +1,210 @@
+        defer_support: false
+        listen: 0.0.0.0:80
+        path: /graphql
+      # We keep subgraph errors on in our own environment, always.
```
Is it worth giving the rationale for this, or just to say:

```suggestion
      # We recommend keeping subgraph errors on in your own environment, always.
```

garypen replied:
We should delete that comment. We are keeping subgraph errors on in-house to provide backwards compatibility with existing Gateway behaviour. It's not ideal, since it may leak implementation detail or PII(Personal Identifiable Information), but it's a requirement of our particular deployment.

Meschreiber commented on docs/shared/router-common-config.mdx:
```diff
@@ -0,0 +1,210 @@
+            endpoint: "${env.NODE_IP}:4317"
+        # Enabling this both includes the `apollo-trace-id` value on the HTTP response,
+        # but also creates "Response Header" nodes in the trace that we send to Studio.
+        # This should allow us to correlate traces from Studio to logs:
```
"should" introduces a note of uncertainty

```suggestion
        # This allows us to correlate traces from Studio to logs:
```

Meschreiber commented on docs/shared/router-common-config.mdx:
```diff
@@ -0,0 +1,210 @@
+  extraVolumes:
+    - name: rhai-volume
+      configMap:
+        # Provide the name of the ConfigMap containing the files you want
```
```suggestion
        # Provide the name of the ConfigMap containing the files
```

Meschreiber commented on docs/source/containerization/docker.mdx:
```diff
@@ -81,11 +113,11 @@ docker run -p 4000:4000 \
 ## Specifying the supergraph
 
-If we don't want to use uplink to retrieve our subgraph, we can manually specify the details.
+If we don't want to use [Apollo Uplink](/federation/managed-federation/uplink/) to retrieve our subgraph, we can manually specify the details.
```
Can we add why anyone would want to do this?

garypen replied:
This is users who don't want to subscribe to our model for automatically updating the supergraph via uplink or who don't have connectivity (air-gapped) for their environment.

Meschreiber reviewed (commented):
Great additions @garypen and @shorgi ! 👏👏👏

I've included inline copy suggestions and have these general remarks:
- Love that you've moved config files to shared content for reusability. Within the content of a page, does it generally make sense to hide them in `ExpansionPanels`? [In some cases](https://deploy-preview-3943--apollo-router-docs.netlify.app/containerization/kubernetes#basic-deployment) you're already doing this, and I'm wondering if there are other places it makes sense too, e.g. https://deploy-preview-3943--apollo-router-docs.netlify.app/containerization/kubernetes#deploy-with-a-coprocessor
- Structure on most pages makes sense. On the Docker page, the intro paragraph doesn't align with section headers:

![image](https://github.com/apollographql/router/assets/23219998/020b3cdd-3f9a-4fa9-9341-a2383552beef)

It would be nice to get them a little more aligned. (E.g., make [Running the debug container to investigate memory issues](https://deploy-preview-3943--apollo-router-docs.netlify.app/containerization/docker#running-the-debug-container-to-investigate-memory-issues) an H3, add [Specifying the supergraph](https://deploy-preview-3943--apollo-router-docs.netlify.app/containerization/docker#specifying-the-supergraph) to the intro.)

Finally, though not (yet) included on this PR, the [Passing command-line arguments to the Router binary](https://deploy-preview-3943--apollo-router-docs.netlify.app/containerization/docker#passing-command-line-arguments-to-the-router-binary) and [Debugging your container](https://deploy-preview-3943--apollo-router-docs.netlify.app/containerization/docker#debugging-your-container) sections could use a little copyediting love too (if time permits).

E.g. The first sentence of [Passing command-line arguments to the Router binary](https://deploy-preview-3943--apollo-router-docs.netlify.app/containerization/docker#passing-command-line-arguments-to-the-router-binary) is a little hard to follow: "By default, the router command invoked inside the published container does run not set any of the [available command-line options](https://deploy-preview-3943--apollo-router-docs.netlify.app/configuration/overview#command-line-options)."

And for [Debugging your container](https://deploy-preview-3943--apollo-router-docs.netlify.app/containerization/docker#debugging-your-container), it would be best to  [avoid using terms like "easy" or "simply" when giving users instructions](https://developers.google.com/style/word-list#easy).

PS. Apologies for starting this review with one-off suggestions instead of incorporating them all here. Was still drinking my morning coffee! 😅

shorgi commented on docs/source/containerization/kubernetes.mdx:
```diff
@@ -1,264 +1,250 @@
+## About the router Helm chart
 
-> Note: The Apollo Router is made available under the Elastic License v2.0 (ELv2).  This applies to its source code and all distributions, including versions installed via Helm charts.  Read [our licensing page](https://www.apollographql.com/docs/resources/elastic-license-v2-faq/) for more details.
+[Helm](https://helm.sh) is a package manager for Kubernetes (k8s). Apollo provides an application Helm chart with each release of Apollo Router in GitHub. Since the router version 0.14.0, Apollo has released the router Helm chart as an [Open Container Initiative (OCI)](https://helm.sh/docs/topics/registries/) image in our GitHub container registry.
```
It's an unfortunate name collision, as GitHub explicitly names it its ["Container registry"](https://docs.github.com/en/packages/working-with-a-github-packages-registry/working-with-the-container-registry). We need an improvement to Apollopedia, to ignore specified instances like this.

Meschreiber replied:
Ah I see, I think it's referred to as "container repository" in a couple of places in different pages in this PR, so it might be worth doing a find/replace.

shorgi commented on docs/source/containerization/kubernetes.mdx:
```diff
@@ -1,264 +1,250 @@
-Starting with release 0.14.0, each time we release the router, we'll release
-our helm chart and store it in the same OCI registry in which we store our
-router docker images.
+* Default container ports for the router's (supergraph) [HTTP server](../../configuration/overview/#listen-address), [health check endpoint](../../configuration/health-checks), and [metrics endpoint](../../configuration/metrics).
```
It's meant to be a callout to the `supergraph` key in the chart's values file. I agree though it's distracting standalone, will instead add a comment in the values file snippet that `supergraph` refers to the HTTP server.

garypen commented on docs/source/config.json:
```diff
@@ -106,8 +106,8 @@
       "Overview": "/containerization/overview",
-      "Docker": "/containerization/docker",
-      "Kubernetes": "/containerization/kubernetes"
+      "Deploy on Kubernetes": "/containerization/kubernetes",
```
Kubernetes tends to be used for production deployments. Docker is more likely to be used by individual developers debugging/testing/etc...

shorgi commented:
> I've included inline copy suggestions and have these general remarks:

Thanks @Meschreiber for your thoughtful and thorough review 🙏 ! I've addressed all of your suggestions and general remarks.

garypen commented on docs/source/containerization/docker.mdx:
````diff
@@ -123,4 +150,4 @@ Usage: build_docker_image.sh [-b [-r <repo>]] [-d] [<release>]
 ```
 
-Note: The script has to be run from the `dockerfiles/diy` directory because it makes assumptions about the relative availability of various files. The example uses [debian:bullseye-slim image](https://hub.docker.com/_/debian/) for the final image build. Feel free to modify the script to use images which better suit your own needs, but be careful if using the [-d] flag since it makes the assumptio ...
+The script has to be run from the `dockerfiles/diy` directory because it makes assumptions about the relative availability of various files. The example uses [debian:bullseye-slim image](https://hub.docker.com/_/debian/) for the final image build. Feel free to modify the script to use images which better suit your own needs, but be careful if using the [-d] flag since it makes the assumption that ...
````
You can delete that first sentence. We fixed the build script a while back so that was no longer required.

shorgi replied:
deleting `The script has to be run ... files.`

garypen commented on docs/source/containerization/kubernetes.mdx:
```diff
@@ -1,264 +1,250 @@
+
+> ⚠️ Your Kubernetes version must be compatible with Helm v3. For details, see [Helm Version Support Policy](https://helm.sh/docs/topics/version_skew/#supported-version-skew).
+
+* Log in with Helm to the Apollo container registry in GitHub.
```
Like @shorgi noted above, this is a "registry" and not a "repository". It will probably cause more confusion if we change this.

garypen commented:
I opened a replacement PR: https://github.com/apollographql/router/pull/4001, so that we can target these changes to main. Closing this one so it doesn't get merged by mistake.
